### PR TITLE
[SPARK-21101][SQL] Catch IllegalStateException when CREATE TEMPORARY FUNCTION

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

It must `override` [`public StructObjectInspector initialize(ObjectInspector[] argOIs)`](https://github.com/apache/hive/blob/release-2.0.0/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTF.java#L70) when create a UDTF. 

If you `override` [`public StructObjectInspector initialize(StructObjectInspector argOIs)`](https://github.com/apache/hive/blob/release-2.0.0/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTF.java#L49), `IllegalStateException` will throw. per: [HIVE-12377](https://issues.apache.org/jira/browse/HIVE-12377).

This PR catch `IllegalStateException` and point user to `override` `public StructObjectInspector initialize(ObjectInspector[] argOIs)`.

## How was this patch tested?

unit tests

Source code and binary jar: [SPARK-21101.zip](https://github.com/apache/spark/files/1123763/SPARK-21101.zip)
These two source code copy from : https://github.com/apache/hive/blob/release-2.0.0/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDTFStack.java

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 1]
```diff
@@ -99,6 +99,14 @@ private[sql] class HiveSessionCatalog(
           throw new AnalysisException(s"No handler for Hive UDF '${clazz.getCanonicalName}'")
         }
       } catch {
+        case ise: IllegalStateException =>
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: Hi, @wangyum .
Can we move this after line 112, `case NonFatal(e) =>`?
IMO, these 8 lines are mostly duplication logic of line 113~116. The only difference is error message.
I guess we can append the new error message more simply.

**Comment by maropu** [comment]
Can you add a test? I think it's okay to add a source file for UDTF in  `sql/hive/src/test/java/org/apache/spark/sql/hive/execution` or `sql/hive/src/test/resources` and add the test.

**Comment by maropu** [comment]
I think it'd be better to fix code so we could call both interfaces (deprecated and non-deprecated `initialize`) instead of throwing the exception. cc: @hvanhovell

**Comment by wangyum** [comment]
Retest this please

**Comment by gatorsmile** [comment]
ping @wangyum This sounds a reasonable fix. Could you resolve the conflicts?

**Comment by wangyum** [comment]
Retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 2]
```diff
@@ -95,7 +95,10 @@ private[sql] class HiveSessionCatalog(
       } catch {
         case NonFatal(e) =>
           val analysisException =
-            new AnalysisException(s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}': $e")
+            new AnalysisException(s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}': $e" +
+              s"\nIf you create a UDTF, please make sure your function override " +
+              s"`public StructObjectInspector initialize(ObjectInspector[] args)`, " +
+              s"per: SPARK-21101")
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]:
```Scala
          val noHandlerMsg = s"No handler for UDF/UDAF/UDTF '${clazz.getCanonicalName}': $e"
          val errorMsg =
            if (classOf[GenericUDTF].isAssignableFrom(clazz)) {
              s"$noHandlerMsg\nPlease make sure your function overrides " +
                "`public StructObjectInspector initialize(ObjectInspector[] args)`."
            } else {
              noHandlerMsg
            }
          val analysisException = new AnalysisException(errorMsg)
```

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

**Pull request closed without merging**